modules/DINO CPU/_posts/2000-05-01-assignment4.md

```diff
@@ -4,7 +4,569 @@ Editor: Hoa Nguyen
+| Part                | Percentage |
+|---------------------|------------|
+| Part 4.1            |        20% |
+| Part 4.2 Question 1 |        10% |
+| Part 4.2 Question 2 |        20% |
+| Part 4.2 Question 3 |        15% |
+| Part 4.2 Question 4 |        15% |
+| Part 4.3 Question 5 |        10% |
+| Part 4.3 Question 6 |        10% |
+| Extra Credits       |        10% |
+
+## Submission
+**Warning:** read the submission instructions carefully. Failure to adhere to the instructions will result in a loss of points.
+
+### Code Portion
+You will upload the following files to Gradescope on the `Assignment 4 - Code Portion` assignment,
+- `src/main/scala/pipelined/cpu-noncombin.scala`
+
+Once uploaded, Gradescope will automatically download and run your code.
+This should take less than 30 minutes.
+For each part of the assignment, you will receive a grade.
+If all of your tests are passing locally, they should also pass on Gradescope unless you made changes to the I/O, which you are not allowed to do.
+
+**Note:** Make sure that you comment out or remove all printing statements in your submissions.
+There are a lot of long tests for this assignment.
+The auto-grader might fail complete grading within the allocated time if there are too many output statements.
+(Outputting to `stdout/stderr` is very costly time-wise!)
+
+### Written Portion
+You will upload your answers for the `Assignment 4 - Written Portion` assignment to Gradescope.
+Please upload a separate page for each answer!
+Additionally, I believe Gradescope allows you to circle the area with your final answer.
+Make sure to do this!
+
+We will not grade any questions for which we are unable to read.
+Be sure to check your submission to make sure it's legible, right-side-up, etc.
+
+### Academic misconduct reminder
+You are to work on this project **individually**.
+You may discuss *high level concepts* with one another (e.g., talking about the diagram), but all work must be completed on your own.
+
+**Remember, DO NOT POST YOUR CODE PUBLICLY ON GITHUB!**
+Any code found on GitHub that is not the base template you are given will be reported to SJA.
+If you want to sidestep this problem entirely, don't create a public fork and instead create a private repository to store your work.
+GitHub now allows everybody to create unlimited private repositories for up to three collaborators, and you shouldn't have any collaborators for your code for this assignment.
+
+## Hints
+- **Start early!** Completing part 4.1 is essential for the next parts.
+- If you need help, come to office hours for the TAs, or post your questions on Piazza.
+
+
+# Extra Credits (10 points)
+
+Improve the performance of the non combination pipelined CPU when running *any* of
+application in the Full Application tests and their `-loops-unrolled` variants.
+You can find the tests [here](https://github.com/ECS154B-WQ23/dinocpu-assignment3/blob/main/src/main/scala/testing/InstTests.scala#L1096)
```

Comment:
Can you use {{154b_github_classroom_link}} here?

Reply:
done